cashu/core/crypto/dlc.py

```diff
@@ -0,0 +1,71 @@
+from hashlib import sha256
+from typing import List, Optional, Tuple
+from secp256k1 import PrivateKey, PublicKey
+
+
+def sorted_merkle_hash(left: bytes, right: bytes) -> bytes:
+    '''Sorts `left` and `right` in non-ascending order and
+        computes the hash of their concatenation
+    '''
+    if left < right:
+        left, right = right, left
+    return sha256(left+right).digest()
+
+
+def merkle_root(
+    leaf_hashes: List[bytes],
+    track_branch: Optional[int] = None
+    ) -> Tuple[bytes, Optional[List[bytes]]]:
+    '''Computes the root of a list of leaf hashes
+        if `track_branch` is set, extracts the hashes for the branch that leads
+        to `leaf_hashes[track_branch]`
+    '''
+    if track_branch is not None:
+        if len(leaf_hashes) == 0:
+            return b"", []
+        elif len(leaf_hashes) == 1:
+            return leaf_hashes[0], []
+        else:
+            split = len(leaf_hashes) // 2
+            left, left_branch_hashes = merkle_root(leaf_hashes[:split],
+                track_branch if track_branch < split else None)
+            right, right_branch_hashes = merkle_root(leaf_hashes[split:],
+                track_branch-split if track_branch >= split else None)
+            branch_hashes = (left_branch_hashes if
+                track_branch < split else right_branch_hashes)
+            hashh = sorted_merkle_hash(left, right)
+            # Needed to pass mypy checks
+            assert branch_hashes is not None, "merkle_root fail: branch_hashes == None"
+            branch_hashes.append(right if track_branch < split else left)
+            return hashh, branch_hashes
+    else:
+        if len(leaf_hashes) == 0:
+            return b"", None
+        elif len(leaf_hashes) == 1:
+            return leaf_hashes[0], None
+        else:
+            split = len(leaf_hashes) // 2
+            left, _  = merkle_root(leaf_hashes[:split], None)
+            right, _ = merkle_root(leaf_hashes[split:], None)
+            hashh = sorted_merkle_hash(left, right)
+            return hashh, None
+
+def merkle_verify(root: bytes, leaf_hash: bytes, proof: List[bytes]) -> bool:
+    '''Verifies that `leaf_hash` belongs to a merkle tree
+        that has `root` as root
+    '''
+    h = leaf_hash
+    for branch_hash in proof:
+        h = sorted_merkle_hash(h, branch_hash)
+    return h == root
+
+def list_hash(leaves: List[str]) -> List[bytes]:
+    return [sha256(leaf.encode()).digest() for leaf in leaves]
+
+def sign_dlc(dlc_root: str, privkey: PrivateKey) -> bytes:
+    dlc_root_hash = sha256(bytes.fromhex(dlc_root)).digest()
+    return privkey.schnorr_sign(dlc_root_hash, None, raw=True)
+
+def verify_dlc_signature(dlc_root: str, signature: bytes, pubkey: PublicKey) -> bool:
+    dlc_root_hash = sha256(bytes.fromhex(dlc_root)).digest()
+    return pubkey.schnorr_verify(dlc_root_hash, signature, None, raw=True)
```

Comment:
I'm starting to think it might be necessary to also have the mint sign the `funding_amount`. Otherwise, the funder could fund the same DLC root hash with a tiny amount, e.g. 1 sat, and use the resulting signature to convince the other participants that the full DLC amount was funded. 

I'll update the spec to reflect this when I have a moment

Reply:
Either works :smile: 

- Matrix: `@conduition`
- Nostr: `npub1l6uy9chxyn943cmylrmukd3uqdq8h623nt2gxfh4rruhdv64zpvsx6zvtg`